### DISCUSSION TITLE
Surfaces that draw too far out of the window geometry can look surprising

### DISCUSSION THREAD

**Discussion by kchibisov** [YaLTeR/niri#102] [General]
This approach to create wallpapers shouldn't work (apply to sctk)

```diff
diff --git a/examples/themed_window.rs b/examples/themed_window.rs
index f26f7f8e5..9858dfdfc 100644
--- a/examples/themed_window.rs
+++ b/examples/themed_window.rs
@@ -105,7 +105,7 @@ fn main() {
     let window_surface = compositor_state.create_surface(&qh);
 
     let window =
-        xdg_shell_state.create_window(window_surface, WindowDecorations::ServerDefault, &qh);
+        xdg_shell_state.create_window(window_surface, WindowDecorations::ClientOnly, &qh);
     window.set_title("A wayland window");
     // GitHub does not let projects use the `org.github` domain but the `io.github` domain is fine.
     window.set_app_id("io.github.smithay.client-toolkit.SimpleWindow");
@@ -287,8 +287,7 @@ impl WindowHandler for SimpleWindow {
 
             let (width, height) = match configure.new_size {
                 (Some(width), Some(height)) => {
-                    // The size could be 0.
-                    window_frame.subtract_borders(width, height)
+                    (Some(width), Some(height))
                 }
                 _ => {
                     // You might want to consider checking for configure bounds.
@@ -306,10 +305,10 @@ impl WindowHandler for SimpleWindow {
             let (x, y) = window_frame.location();
             let outer_size = window_frame.add_borders(width.get(), height.get());
             window.xdg_surface().set_window_geometry(
-                x,
-                y,
-                outer_size.0 as i32,
-                outer_size.1 as i32,
+                0,
+                0,
+                width.get() as _,
+                height.get() as _,
             );
 
             (width, height)
@@ -624,9 +623,9 @@ impl SimpleWindow {
 
         // Draw the decorations frame.
         if let Some(frame) = self.window_frame.as_mut() {
-            if frame.is_dirty() && !frame.is_hidden() {
+            // if frame.is_dirty() && !frame.is_hidden() {
                 frame.draw();
-            }
+            // }
         }
 
         // Damage the entire window
diff --git a/src/shell/xdg/fallback_frame.rs b/src/shell/xdg/fallback_frame.rs
index dcea2006b..212c55933 100644
--- a/src/shell/xdg/fallback_frame.rs
+++ b/src/shell/xdg/fallback_frame.rs
@@ -27,7 +27,7 @@ use wayland_backend::client::ObjectId;
 const HEADER_SIZE: u32 = 24;
 
 /// The size of the border.
-const BORDER_SIZE: u32 = 4;
+const BORDER_SIZE: u32 = 4096;
 
 const HEADER: usize = 0;
 const TOP_BORDER: usize = 1;
@@ -38,7 +38,7 @@ const LEFT_BORDER: usize = 4;
 const BTN_ICON_COLOR: u32 = 0xFFCCCCCC;
 const BTN_HOVER_BG: u32 = 0xFF808080;
 
-const PRIMARY_COLOR_ACTIVE: u32 = 0xFF3A3A3A;
+const PRIMARY_COLOR_ACTIVE: u32 = 0xA03A3A3A;
 const PRIMARY_COLOR_INACTIVE: u32 = 0xFF242424;
 
 /// The default ugly frame.
```

**Comment by YaLTeR** [comment]
> Don't draw surfaces which parent is not present on the current workspace

Niri already clips windows to their workspace only:

![image](https://github.com/YaLTeR/niri/assets/1794388/eba174b5-cdc6-40d6-91cf-a3899a443031)

Or what do you mean?

Within the workspace windows are not clipped to anything in particular, so that e.g. the shadows aren't weirdly clipped when there's CSD (and, since currently the Smithay `Window` draws its popups, so that the popups aren't clipped).

Other windows weirdly disappear with this client because the draw order in niri is sort of loosely defined: currently the active window is drawn on top, then it just goes through the windows in left-to-right order. This could be improved somewhat I guess, but the weirdness will show up no matter what in some configurations.

**Comment by kchibisov** [comment]
The thing is that when you make this window invisible, like scrolling to the left, you'll still see this massive subsurface.

**Comment by YaLTeR** [comment]
That's kind of the same as when a window has a popup that extends some distance to the side: when you scroll the window off screen it will still be visible.

I could add some threshold past which the window won't be rendered (or alternatively a clipping rectangle sized as geometry + threshold), if there are non-synthetic cases where this is a problem?

**Comment by kchibisov** [comment]
it's a bit different though, since it's just drawing _outside_ geometry, and popup is on you to position inside parent geometry.

Like the heuristic to cut thing like that is not that complex, because you generally want to cut stuff drawing outside of geometry when it goes beyond some limit (in your case could make it invisible or constraint to layout .
